api/src/main/java/org/openmrs/api/context/UserContext.java

```diff
@@ -330,7 +331,7 @@ public Set<Role> getAllRoles(User user) throws Exception {
 	public boolean hasPrivilege(String privilege) {
 		log.debug("Checking '{}' against proxies: {}", privilege, proxies);
 		// check proxied privileges
-		for (String s : new ArrayList<>(proxies)) {
```

Comment:
I'm running into a java.util.ConcurrentModificationException with this change. This is also causing some issues with the legacy UI.

```
java.util.ConcurrentModificationException
	at java.util.ArrayList$Itr.checkForComodification(ArrayList.java:911)
	at java.util.ArrayList$Itr.next(ArrayList.java:861)
	at org.openmrs.api.context.UserContext.hasPrivilege(UserContext.java:334)
	at org.openmrs.api.context.Context.hasPrivilege(Context.java:738)
	at org.openmrs.aop.AuthorizationAdvice.before(AuthorizationAdvice.java:78)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:57)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:241)
	at com.sun.proxy.$Proxy281.getGlobalProperty(Unknown Source)
	at org.openmrs.module.owa.filter.OwaFilter.doFilter(OwaFilter.java:53)
	at org.openmrs.module.web.filter.ModuleFilterChain.doFilter(ModuleFilterChain.java:71)
	at org.openmrs.module.web.filter.ModuleFilter.doFilter(ModuleFilter.java:57)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.owasp.csrfguard.CsrfGuardFilter.handleSession(CsrfGuardFilter.java:107)
	at org.owasp.csrfguard.CsrfGuardFilter.doFilter(CsrfGuardFilter.java:97)
	at org.owasp.csrfguard.CsrfGuardFilter.doFilter(CsrfGuardFilter.java:68)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.openmrs.web.filter.OpenmrsFilter.doFilterInternal(OpenmrsFilter.java:114)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.openmrs.web.filter.CookieClearingFilter.doFilterInternal(CookieClearingFilter.java:77)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.springframework.orm.hibernate5.support.OpenSessionInViewFilter.doFilterInternal(OpenSessionInViewFilter.java:156)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.springframework.web.multipart.support.MultipartFilter.doFilterInternal(MultipartFilter.java:125)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.openmrs.web.filter.StartupFilter.doFilter(StartupFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.openmrs.web.filter.StartupFilter.doFilter(StartupFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.openmrs.web.filter.StartupFilter.doFilter(StartupFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:117)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:168)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:144)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:168)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:482)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:130)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:346)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:383)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:937)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1791)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
	at java.lang.Thread.run(Thread.java:750)
```

![Screenshot from 2024-10-05 10-38-01](https://github.com/user-attachments/assets/8f02f2ce-5617-46ee-b8e2-1e9907f241ef)

Reply:
@wikumChamith Ahh... my assumption is wrong then. It's likely possible that the `proxies` array is modified within the loop. I'll keep this change. Thanks for testing.

Reply:
Yeah, the synchronized list does not actually synchronize the iterator. I was hoping we'd have a more comprehensive solution for this, but I guess if it works, it works.

Reply:
That is exactly why i was hesitant to merge this in. But if you can confirm that it fixes the reported problem, then that is good. 😊

Reply:
Well, it doesn't harm anything to merge this in, but I take your point.  @chibongho , @mogoodrich and I just talked and the approach we plan to take is to backport the additional fix that wraps proxies in a new List before iterating it from master, since it is already committed there and this would make the 2 branches consistent.  This seems like it solves this issue.

That said, the _right_ solution is likely different, as @ibacher suggests.  I would think that changing "proxies" to a ThreadLocal might be the most correct solution.  Proxy privileges afaik, should always be added and removed for a very specific task within a Thread, not something that persists between threads.  @dkayiwa ?

Reply:
Yeah, I think a ThreadLocal is more likely to be the right solution. The one thing that would break is that a child thread wouldn't inherit the proxies from it's parent thread, but I don't think that's something we've intended to guarantee.

Reply:
Two things:

1. I can't reproduce this issue at all with 2.6.8. It showed up in master, but not 2.6.8 (test3.openmrs.org is running 2.6.8).
2. The weird thing about ThreadLocal is that [the UserContext itself](https://github.com/openmrs/openmrs-core/blob/8d46a47403fbc66852fe447011b3750d196a5257/api/src/main/java/org/openmrs/api/context/Context.java#L148) is a ThreadLocal and I'm not sure how a single thread could produce contention like this (since we don't AFAIK use any kind of "green thread" setup).